source/extensions/bootstrap/wasm/config.h

```diff
@@ -19,9 +19,21 @@ namespace Wasm {
 class WasmService {
 public:
   WasmService(Common::Wasm::WasmHandleSharedPtr singleton) : singleton_(std::move(singleton)) {}
-  WasmService(ThreadLocal::SlotPtr tls_slot) : tls_slot_(std::move(tls_slot)) {}
+  WasmService(Common::Wasm::PluginSharedPtr plugin, ThreadLocal::SlotPtr tls_slot)
+      : plugin_(plugin), tls_slot_(std::move(tls_slot)) {}
+
+  virtual ~WasmService() {
+    if (tls_slot_ && tls_slot_->currentThreadRegistered() && tls_slot_->get()) {
```

Comment:
Same comment here about registration. Seems like we have some strange lifetime issues going on.

Reply:
I honestly have no idea, but without this check, this assert fails during shutdown in debug builds:
```
[critical][assert] [source/common/thread_local/thread_local_impl.cc:159] assert failure: !shutdown_.                                                                                                                         
```
Any ideas, @jmarantz?

Reply:
I'm not saying that it's a result of your recent changes, but you're probably most familiar with that code.